<a href="https://colab.research.google.com/github/MuthiahAinun/Proyek_Analisis_Sentimen/blob/main/Skema2_Proyek_Analisis_Sentimen_Tsamarah_Muthi'ah_A.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Proyek Analisis Sentimen: [Youtube-Comments-dataset]
- **Nama:** [Tsamarah Muthi'ah Abdullah]
- **Email:** [a135xaf486@devacademy.id]
- **ID Dicoding:** [a135xaf48]

# Import Library

In [ ]:
# Menginstall google API
!pip install --upgrade google-api-python-client pandas

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 89.9/89.9 kB 2.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.1/13.1 MB 75.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.1/13.1 MB 78.4 MB/s eta 0:00:00
  Attempting uninstall: pandas
    Found existing installation: pandas 2.2.2
    Uninstalling pandas-2.2.2:
      Successfully uninstalled pandas-2.2.2
  Attempting uninstall: google-api-python-client
    Found existing installation: google-api-python-client 2.160.0
    Uninstalling google-api-python-client-2.160.0:
      Successfully uninstalled google-api-python-client-2.160.0
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
google-colab 1.0.0 requires pandas==2.2.2, but you have pandas 2.2.3 which is incompatible.


In [3]:
!pip install gensim

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.0/61.0 kB 2.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.6/60.6 kB 2.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 26.7/26.7 MB 46.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 18.3/18.3 MB 81.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 38.6/38.6 MB 15.6 MB/s eta 0:00:00
  Attempting uninstall: numpy
    Found existing installation: numpy 2.0.2
    Uninstalling numpy-2.0.2:
      Successfully uninstalled numpy-2.0.2
  Attempting uninstall: scipy
    Found existing installation: scipy 1.14.1
    Uninstalling scipy-1.14.1:
      Successfully uninstalled scipy-1.14.1


In [7]:
!pip install --upgrade numpy

In [10]:
!pip install --upgrade gensim

In [1]:
# Import Library
import googleapiclient.discovery
import pandas as pd
import time
import random
import re
import string
import tensorflow as tf
import numpy as np
from tensorflow.keras.layers import Embedding, LSTM, Dense, Dropout, Conv1D, GlobalMaxPooling1D, Bidirectional, BatchNormalization
from sklearn.metrics import classification_report, accuracy_score
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, LSTM, GRU, Conv1D, GlobalMaxPooling1D, Dense, Dropout, Bidirectional, BatchNormalization
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from imblearn.over_sampling import SMOTE
from gensim.models import Word2Vec

# Data Preprocessing dan Pembersihan

In [4]:
# Membaca dan Menampilkan informasi dataset
dataset = pd.read_csv("NDsO1LT_0lw_youtube_comments.csv")
print("Jumlah data:", len(dataset))

Jumlah data: 85028


In [3]:
print(dataset.head())

             username                                               text  \
0            @MrBeast  BEAST GAMES FINALE DROPS FEBRUARY 13TH! GO WAT...   
1     @SultanRayyan-7                          Saya orang Indonesia 🇮🇩🇮🇩   
2  @MohamadrezaRezayy  مستر بیست عزیز ❤ لطفا دوبله فارسی این ویدیو را...   
3       @ILoveyou-954  বাংলাদেশ আসবেন কবে পাবনা জেলার কিছু দেখতে চাই ...   
4     @Moneymakerarab  This is for girls listen to Dalida - Helwa ya ...   

   likes          published_at  
0  54600  2025-02-08T16:59:31Z  
1      0  2025-03-19T21:51:42Z  
2      0  2025-03-19T21:43:03Z  
3      0  2025-03-19T21:34:32Z  
4      0  2025-03-19T21:20:21Z  


In [5]:
print("Informasi Dataset:")
print(dataset.info())

Informasi Dataset:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85028 entries, 0 to 85027
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   username      84982 non-null  object
 1   text          85028 non-null  object
 2   likes         85028 non-null  int64 
 3   published_at  85028 non-null  object
dtypes: int64(1), object(3)
memory usage: 2.6+ MB
None


In [6]:
print('\nStatistik Deskriptif:')
print(dataset.describe())


Statistik Deskriptif:
               likes
count   85028.000000
mean        9.727654
std       744.702565
min         0.000000
25%         0.000000
50%         0.000000
75%         1.000000
max    167348.000000


In [7]:
# Mengecek jumlah missing values di setiap kolom
print(dataset.isnull().sum())

username        46
text             0
likes            0
published_at     0
dtype: int64


In [8]:
# Menghapus nilai kosong
dataset.dropna(inplace=True)

In [9]:
# Menghapus duplikasi
dataset.drop_duplicates(subset=['text'], inplace=True)

In [10]:
# Pembersihan teks
def clean_text(text):
    text = text.lower()
    text = ''.join(c for c in text if c.isalnum() or c.isspace())
    return text

dataset['cleaned_text'] = dataset['text'].apply(clean_text)
print("Data setelah pembersihan:")
print(dataset.head())

Data setelah pembersihan:
             username                                               text  \
0            @MrBeast  BEAST GAMES FINALE DROPS FEBRUARY 13TH! GO WAT...   
1     @SultanRayyan-7                          Saya orang Indonesia 🇮🇩🇮🇩   
2  @MohamadrezaRezayy  مستر بیست عزیز ❤ لطفا دوبله فارسی این ویدیو را...   
3       @ILoveyou-954  বাংলাদেশ আসবেন কবে পাবনা জেলার কিছু দেখতে চাই ...   
4     @Moneymakerarab  This is for girls listen to Dalida - Helwa ya ...   

   likes          published_at  \
0  54600  2025-02-08T16:59:31Z   
1      0  2025-03-19T21:51:42Z   
2      0  2025-03-19T21:43:03Z   
3      0  2025-03-19T21:34:32Z   
4      0  2025-03-19T21:20:21Z   

                                        cleaned_text  
0  beast games finale drops february 13th go watc...  
1                              saya orang indonesia   
2  مستر بیست عزیز  لطفا دوبله فارسی این ویدیو را ...  
3                    বলদশ আসবন কব পবন জলর কছ দখত চই   
4  this is for girls listen to dalida

In [11]:
# Mengecek kembali jumlah missing values di setiap kolom
print(dataset.isnull().sum())

username        0
text            0
likes           0
published_at    0
cleaned_text    0
dtype: int64


# **2. Ekstraksi Fitur dan Pelabelan Data**

Metode yang digunakan bebas sesuai dengan preferensi masing-masing peserta. Tahapan ini penting untuk mempersiapkan data sehingga dapat diolah lebih lanjut dalam proses pelatihan model.

In [12]:
# Melabeli komentar secara manual (positif, netral, negatif)
def label_comment(text):
    if any(word in text for word in ["love", "great", "awesome", "good", "nice", "amazing", "fantastic"]):
        return "positif"
    elif any(word in text for word in ["hate", "bad", "awful", "worst", "terrible", "disgusting", "boring"]):
        return "negatif"
    else:
        return "netral"

dataset['label'] = dataset['cleaned_text'].apply(label_comment)
print("Distribusi label:")
print(dataset['label'].value_counts())

Distribusi label:
label
netral     63766
positif     4728
negatif       69
Name: count, dtype: int64


In [13]:
# Encoding label
label_encoder = LabelEncoder()
dataset['label_encoded'] = label_encoder.fit_transform(dataset['label'])

In [16]:
# Tokenizer dan Padding
tokenizer = Tokenizer(num_words=10000, oov_token="<OOV>")
tokenizer.fit_on_texts(dataset['cleaned_text'])
sequences = tokenizer.texts_to_sequences(dataset['cleaned_text'])
padded = pad_sequences(sequences, maxlen=100, padding='post')

In [17]:
# Mengatasi Ketidakseimbangan Data dengan SMOTE
smote = SMOTE(random_state=42)
X_resampled, y_resampled = smote.fit_resample(padded, dataset['label_encoded'])

# **3. Pembangunan Model Deep Learning**

Pilihan algoritma pelatihan ini haruslah sesuai dengan tujuan analisis sentimen yang ingin dicapai.

# Skema 2: CNN dengan Word2Vec (80/20)

**Algoritma:** Convolutional Neural Network (CNN)

**Ekstraksi Fitur:** Word2Vec

**Alasan Pemilihan:**
- CNN dikenal dalam pengenalan pola spasial, dan bisa dimanfaatkan untuk menangkap pola lokal pada teks (misalnya n-grams).
- Word2Vec menghasilkan representasi vektor dari kata-kata dengan mempertimbangkan hubungan semantik dan sintaksis.
- Kombinasi CNN + Word2Vec memungkinkan model mempelajari fitur spasial dari representasi kata yang lebih kaya.

In [18]:
# Membagi data menjadi training dan testing
X_train, X_test, y_train, y_test = train_test_split(X_resampled, y_resampled, test_size=0.2, random_state=42)

In [19]:
# Membuat model Word2Vec menggunakan teks bersih
word2vec = Word2Vec(sentences=[text.split() for text in dataset['cleaned_text']], vector_size=100, window=5, min_count=1)

In [22]:
# Menyiapkan embedding matrix
embedding_matrix = np.zeros((len(tokenizer.word_index) + 1, 100))
for word, i in tokenizer.word_index.items():
    if word in word2vec.wv:
        embedding_matrix[i] = word2vec.wv[word]

In [23]:
# Membangun model CNN
model2 = Sequential([
    Embedding(len(tokenizer.word_index) + 1, 100, weights=[embedding_matrix], input_length=100, trainable=False),
    Conv1D(128, 5, activation='relu'),
    GlobalMaxPooling1D(),
    Dense(64, activation='relu'),
    Dropout(0.5),
    Dense(3, activation='softmax')
])

In [24]:
model2.compile(loss='sparse_categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

In [34]:
model2.fit(X_train, y_train, epochs=60, validation_data=(X_test, y_test), batch_size=64)

Epoch 1/60
2392/2392 ━━━━━━━━━━━━━━━━━━━━ 108s 45ms/step - accuracy: 0.9189 - loss: 0.2058 - val_accuracy: 0.9087 - val_loss: 0.2575
Epoch 2/60
2392/2392 ━━━━━━━━━━━━━━━━━━━━ 107s 45ms/step - accuracy: 0.9204 - loss: 0.2038 - val_accuracy: 0.9056 - val_loss: 0.2664
Epoch 3/60
2392/2392 ━━━━━━━━━━━━━━━━━━━━ 145s 46ms/step - accuracy: 0.9185 - loss: 0.2075 - val_accuracy: 0.9114 - val_loss: 0.2621
Epoch 4/60
2392/2392 ━━━━━━━━━━━━━━━━━━━━ 141s 46ms/step - accuracy: 0.9190 - loss: 0.2045 - val_accuracy: 0.9107 - val_loss: 0.2479
Epoch 5/60
2392/2392 ━━━━━━━━━━━━━━━━━━━━ 142s 46ms/step - accuracy: 0.9209 - loss: 0.2009 - val_accuracy: 0.9092 - val_loss: 0.2787
Epoch 6/60
2392/2392 ━━━━━━━━━━━━━━━━━━━━ 108s 45ms/step - accuracy: 0.9206 - loss: 0.2013 - val_accuracy: 0.9094 - val_loss: 0.2597
Epoch 7/60
2392/2392 ━━━━━━━━━━━━━━━━━━━━ 142s 45ms/step - accuracy: 0.9194 - loss: 0.2045 - val_accuracy: 0.9063 - val_loss: 0.2656
Epoch 8/60
2392/2392 ━━━━━━━━━━━━━━━━━━━━ 143s 46ms/step - accuracy: 

In [35]:
# Evaluasi model pada data testing
test_loss, test_accuracy = model2.evaluate(X_test, y_test)
print(f"Akurasi Testing (dari model.evaluate): {test_accuracy:.2%}")

1196/1196 ━━━━━━━━━━━━━━━━━━━━ 12s 10ms/step - accuracy: 0.9165 - loss: 0.2599
Akurasi Testing (dari model.evaluate): 91.41%


In [36]:
# Inference dan Testing Skema 2
print("\nTesting Skema 2: CNN dengan Word2Vec (80/20)")
y_pred2 = np.argmax(model2.predict(X_test), axis=1)
print(classification_report(y_test, y_pred2, target_names=label_encoder.classes_))


Testing Skema 2: CNN dengan Word2Vec (80/20)
1196/1196 ━━━━━━━━━━━━━━━━━━━━ 10s 9ms/step
              precision    recall  f1-score   support

     negatif       0.90      0.89      0.89     12749
      netral       0.92      0.97      0.94     12623
     positif       0.93      0.89      0.91     12888

    accuracy                           0.91     38260
   macro avg       0.91      0.91      0.91     38260
weighted avg       0.91      0.91      0.91     38260



**Insight Skema 2 :**

- Model CNN dengan Word2Vec mencapai akurasi 91.41% pada data uji. Angka ini menunjukkan bahwa model cukup andal dalam mengklasifikasikan sentimen komentar YouTube dengan tingkat kesalahan sekitar 8.59%.
- Precision: Persentase prediksi benar dari semua prediksi pada kelas tertentu.

 **Precision tertinggi:** Kelas positif (0.93), menunjukkan model paling akurat dalam mengenali komentar positif.
- Recall: Persentase data aktual yang teridentifikasi benar.

 **Recall tertinggi:** Kelas netral (0.97), artinya hampir semua komentar netral berhasil dideteksi.
- F1-Score: Rata-rata harmonik dari precision dan recall.

 **F1-Score tertinggi**: Kelas netral (0.94), menunjukkan keseimbangan yang baik antara presisi dan recall.

- **Macro Average:** Akurasi pada semua kelas seimbang (0.91), menunjukkan performa model yang konsisten.
- **Weighted Average:** Nilainya juga 0.91, menunjukkan tidak ada kelas yang terlalu mendominasi atau terlalu lemah.